hooks/shotgun_launch_publish.py

```diff
@@ -77,37 +78,119 @@ def execute(self, path, context, associated_entity, **kwargs):
+            bootstrap.
+
+        :raises: RuntimeError when a usable launcher isn't found.
         """
-        Tries to create folders then launch the publish.
+        engine = self.parent.engine
+        launchapp_system_name = "tk-multi-launchapp"
+
+        # The apps are keyed by app instance name. We don't actually care
+        # what the specific instance is called, as we just want some instance
+        # of tk-multi-launchapp. Because of that, we need to check more than
+        # just whether the name of the app is a key in the engine's apps
+        # property.
+        launchapp_commands = []
+
+        for command_name, command_data in engine.commands.iteritems():
+            props = command_data["properties"]
+            app = props.get("app")
+            if app is not None and app.name == launchapp_system_name:
+                if props.get("engine_name") == engine_instance_name:
+                    launchapp_commands.append(command_data)
+
+        if not launchapp_commands:
+            raise RuntimeError(
+                "Unable to find an instance of %s currently running!" % launchapp_system_name
+            )
+
+        # Check to see if there's a group default. Use that if there is, and if
+        # there isn't then we take the first one off the top.
+        launch_callback = launchapp_commands[0]["callback"]
+
+        for command_data in launchapp_commands:
+            if command_data["properties"].get("group_default"):
+                launch_callback = command_data["callback"]
+                break
+
+        launch_callback(file_to_open=path)
+
+    def _get_legacy_launch_command(self, launch_app_instance_name):
+        """
+        Looks for a legacy-style launcher in the current list of running apps.
+
+        :param str launch_app_instance_name: The name of the launchapp instance to look for.
+
+        :returns: A launcher app instance, or None.
         """
-        
         # in older configs, launch instances were named tk-shotgun-launchmaya
         # in newer configs, launch instances are named tk-multi-launchamaya
         old_config = "tk-shotgun-%s" % launch_app_instance_name
         new_config = "tk-multi-%s" % launch_app_instance_name
+        app_instance = None
         
         if old_config in self.parent.engine.apps:
             # we have a tk-shotgun-xxx instance in the config
             app_instance = old_config
-        
         elif new_config in self.parent.engine.apps:
             # we have a tk-multi-xxx instance in the config
             app_instance = new_config
-            
-        else:
-            raise TankError("The '%s' app could not be found in the '%s' "
-                            "environment!" % (new_config, self.parent.engine.environment.get("name")))        
-        
-        
+
+        return app_instance
+
+    def _do_launch(self, launch_app_instance_name, engine_name, path, context):
```

Comment:
minor: suggest moving this just after the `execute` method since it is what is called from there.